samples/hello-oboe/src/main/cpp/PlayAudioEngine.cpp

```diff
@@ -21,202 +21,31 @@
- * after a stream has started because the timestamps are not yet available.
- */
-oboe::Result
-PlayAudioEngine::calculateCurrentOutputLatencyMillis(oboe::AudioStream *stream,
-                                                     double *latencyMillis) {
-
-    // Get the time that a known audio frame was presented for playing
-    auto result = stream->getTimestamp(CLOCK_MONOTONIC);
-
-    if (result == oboe::Result::OK) {
-
-        oboe::FrameTimestamp playedFrame = result.value();
-
-        // Get the write index for the next audio frame
-        int64_t writeIndex = stream->getFramesWritten();
-
-        // Calculate the number of frames between our known frame and the write index
-        int64_t frameIndexDelta = writeIndex - playedFrame.position;
-
-        // Calculate the time which the next frame will be presented
-        int64_t frameTimeDelta = (frameIndexDelta * oboe::kNanosPerSecond) /  (stream->getSampleRate());
-        int64_t nextFramePresentationTime = playedFrame.timestamp + frameTimeDelta;
-
-        // Assume that the next frame will be written at the current time
-        using namespace std::chrono;
-        int64_t nextFrameWriteTime =
-                duration_cast<nanoseconds>(steady_clock::now().time_since_epoch()).count();
-
-        // Calculate the latency
-        *latencyMillis = static_cast<double>(nextFramePresentationTime - nextFrameWriteTime)
-                         / kNanosPerMillisecond;
-    } else {
-        LOGE("Error calculating latency: %s", oboe::convertToText(result.error()));
-    }
-
-    return result;
-}
-
-/**
- * If there is an error with a stream this function will be called. A common example of an error
- * is when an audio device (such as headphones) is disconnected. It is safe to restart the stream
- * in this method. There is no need to create a new thread.
- *
- * @param audioStream the stream with the error
- * @param error the error which occured, a human readable string can be obtained using
- * oboe::convertToText(error);
- *
- * @see oboe::StreamCallback
- */
-void PlayAudioEngine::onErrorAfterClose(oboe::AudioStream *oboeStream, oboe::Result error) {
-    if (error == oboe::Result::ErrorDisconnected) {
-        oboe::AudioStreamBuilder builder_ = oboe::AudioStreamBuilder(*oboeStream);
-        restartStream(&builder_);
-    }
-}
-
-void PlayAudioEngine::restartStream(oboe::AudioStreamBuilder *builder) {
-    LOGI("Restarting stream");
-    createPlaybackStream(builder);
+    mIsLatencyDetectionSupported = (mStream->getTimestamp((CLOCK_MONOTONIC)) !=
+            oboe::Result::ErrorUnimplemented);
+    getCallbackPtr()->setLatencyDetectionEnabled(mIsLatencyDetectionSupported);
+    getCallbackPtr()->setOutputLatencyMillis(&mCurrentOutputLatencyMillis);
+    mLatencyTuner = std::make_unique<oboe::LatencyTuner>(*mStream);
+    getCallbackPtr()->setLatencyTuner(mLatencyTuner.get());
 }
 
 double PlayAudioEngine::getCurrentOutputLatencyMillis() {
```

Comment:
See above.

Reply:
Although it may be a good idea to add a `getLatency` method to Oboe streams, we want ManagedStream to be zero overhead relative to AudioStreams (rather than the two classes diverging in functionality and performance).